karapace/kafka_rest_apis/__init__.py

```diff
@@ -40,36 +50,70 @@ def __init__(self, config: Config) -> None:
         super().__init__(config=config)
         self._add_kafka_rest_routes()
         self.serializer = SchemaRegistrySerializer(config=config)
-        self._cluster_metadata = None
-        self._metadata_birth = None
-        self.metadata_max_age = self.config["admin_metadata_max_age"]
-        self.admin_client = None
-        self.admin_lock = None
-        self.metadata_cache = None
-        self.schemas_cache = {}
-        self.consumer_manager = ConsumerManager(config=config)
-        self.init_admin_client()
+        self.proxies: Dict[str, "UserRestProxy"] = {}
+        self._proxy_lock = asyncio.Lock()
+        log.info("REST proxy starting with (delegated authorization=%s)", self.config.get("rest_authorization", False))
+        self._idle_proxy_janitor_task: Optional[asyncio.Task] = None
 
-        self._async_producer_lock = None
-        self._async_producer: Optional[AIOKafkaProducer] = None
+    async def close(self) -> None:
+        if self._idle_proxy_janitor_task is not None:
+            self._idle_proxy_janitor_task.cancel()
+            self._idle_proxy_janitor_task = None
+        async with AsyncExitStack() as stack:
+            stack.push_async_callback(super().close)
+            stack.push_async_callback(self.serializer.close)
 
-        self.app.on_startup.append(self.initialize_asyncio_locks)
+            for proxy in self.proxies.values():
+                stack.push_async_callback(proxy.aclose)
 
-    async def initialize_asyncio_locks(self, app: aiohttp.web.Application):  # pylint: disable=unused-argument
-        """The locks can be initialized when asyncio loop is available"""
-        self.admin_lock = asyncio.Lock()
-        self._async_producer_lock = asyncio.Lock()
+    async def _idle_proxy_janitor(self) -> None:
```

Comment:
Comments for `_idle_proxy_janitor`

* There are no tests nor other prove that this works as expected. In prod it will be hard to see if this is working as expected, since there's no user-visible error or something like that that would occur if this doesn't work.
* As a reviewer the code is not simple to read and reason about, there's nested `while`, `async with`, `for`, `if`, another `if`and inside there a break. Very  complex, and especially since there are no tests, hard to say whether this works or not.
* Two fixes available. 1) simplify the code so that it's easier to read and/or 2) add tests. Preferably both.
* I think the method could be split into two - 1) a function that gathers the key,proxy-tuples and 2) the delete close of these . The function can be pulled out from the `KafkaRest` class and unit tested.
* There's no reason except for potentially simpler code for "Always clean one at time". For the reader of the code it's unclear _why_ clean only one. Now this makes the reader wonder, is there a reason for this. Let's just clean all that need to be cleaned. That's what would be expected for such a cleaner.